src/app/features/worklog/worklog-export/worklog-export.util.ts

```diff
@@ -4,190 +4,188 @@ import { msToString } from '../../../ui/duration/ms-to-string.pipe';
         }
-        groups[groupKey].timeEstimate += taskGroups[groupKey].timeEstimate;
-        groups[groupKey].timeSpent += taskGroups[groupKey].timeSpent;
-      } else {
-        groups[groupKey] = taskGroups[groupKey];
+        taskGroups[day].timeEstimate += rowItem.timeEstimate;
+        taskGroups[day].timeSpent += rowItem.timeSpent;
       }
     });
+  }
+  return taskGroups;
+};
 
-  });
-
-  const rows: RowItem[] = [];
-  Object.keys(groups).sort().forEach((key => {
-    const group = groups[key];
+/**
+ * If we're grouping by parent task ignore subtasks
+ * If we're grouping by task ignore parent tasks
+ */
+const skipTask = (task: WorklogTask, groupBy: WorklogGrouping): boolean => {
+  return (groupBy === WorklogGrouping.PARENT && task.parentId !== null)
+  || (groupBy === WorklogGrouping.TASK && task.subTaskIds.length > 0);
+};
 
-    group.titlesWithSub = unique(group.tasks.map(t => t.title));
-    // TODO check all the typing
-    group.titles = unique<any>(
-      group.tasks.map((t: WorklogTask) => (
-        t.parentId && (tasks.find(ptIN => ptIN.id === t.parentId) as WorklogTask).title
-      ) || (!t.parentId && t.title)
-      )
-    ).filter((title: string) => !!title);
-    group.dates = unique(group.dates).sort((a: string, b: string) => {
-      const dateA: number = new Date(a).getTime();
-      const dateB: number = new Date(b).getTime();
-      if (dateA === dateB) {
-        return 0;
-      } else if (dateA < dateB) {
-        return -1;
-      }
-      return 1;
-    });
+/**
+ * For each task creates a new rowItem without needing to push to previous taskGroups, unlike handleDateGroup.
+ * We're still creating a map since we will use the key for sorting in the next step.
+ */
+const handleTaskGroup = (data: WorklogExportData, groupBy: WorklogGrouping): ItemsByKey<RowItem> => {
+  const taskGroups: ItemsByKey<RowItem> = {};
+  for (const task of data.tasks) {
+    if (skipTask(task, groupBy)) { continue; };
+    const taskFields = getTaskFields(task, data);
+    const dates = sortDateStrings(Object.keys(task.timeSpentOnDay));
+    taskGroups[task.id] = {
+      dates,
+      timeEstimate: task.timeEstimate,
+      timeSpent: Object.values(task.timeSpentOnDay).reduce((acc, curr) => acc + curr, 0),
+      workStart: 0,
+      workEnd: 0,
+      ...taskFields
+    };
+  }
+  return taskGroups;
+};
 
-    group.notes = group.notes.map((note) => {
-      return note.replace(/\n/g, ' - ');
+/**
+ * For each task creates a new rowItem without needing to push to previous taskGroups, unlike handleDateGroup.
+ * We're still creating a map since we will use the key for sorting in the next step.
+ */
+const handleWorklogGroup = (data: WorklogExportData): ItemsByKey<RowItem> => {
+  const taskGroups: ItemsByKey<RowItem> = {};
+  for (const task of data.tasks) {
+    Object.keys(task.timeSpentOnDay).forEach(day => {
+      const groupKey = task.id + '_' + day;
```

Comment:
Changing this to: `const groupKey = day + '_' + task.id;` would cause the worklog group to be sorted by date. I don't think it makes much sense as it is now.

Reply:
Makes sense to me!